Feature/#142 auth doc (#146)
Avaiga/taipy-doc: docs/manuals/auth/index.md

> # Taipy's Authentication and Authorization

Reviewer: ```suggestion
```

> !!! warning "Reminder: Authorization and Authentication is only available in enterprise version"

Reviewer 2: "in enterprise version"
->
"in the Enterprise version"

> ## Pre-requisite

Reviewer 2: Pre-requisite**s**

> ## Pre-requisite

Reviewer: ```suggestion
# Pre-requisite
```

> An LDAP server must be setup by the user. Taipy will not be responsible for setting the server up.

Reviewer: Should we also specify that the LDAP server should be populated with users with roles?

Author: you mean the ldap provided must have roles similar to our pre-defined roles? i totally agree with that, just want to make sure it's the case :D

> An LDAP server must be setup by the user. Taipy will not be responsible for setting the server up.

Reviewer 2: "must be setup" -> "must be set up"
"Taipy will not be responsible." -> "Taipy provides no support for setting up the server."

> ## Config

Reviewer 2: Configuration

> ## Config

Reviewer: ```suggestion
# Config
```

> Set up connection to your LDAP server with Taipy Config. Here are the configurable properties you can use:

Reviewer 2: "Set up connection to your LDAP server with Taipy Config"
->
"You must configure how Taipy connects to the LDAP server using the Taipy \`Config^\` features"

> - **Config.global_config.ldap_server**: The url of the ldap server.

Reviewer 2: ldap -> LDAP

- **Config.global_config.ldap_base_dn**: The base DN that the LDAP server uses to search for users authentication.

> - **Config.global_config.jwt_secret**: The secret for the JWT token.

Reviewer 2: Please introduce that TaipyAuth uses JWT at some point. People don't know anything about that.
You may even want to drop the 'JWT' part, so you don't have to explain: the more I think about it, the less I see a need to expose the JWT thing. Please just refer to 'a secret' that programmers can use, and expiration period.

Author: I will keep it as 'The secret that will be used to generate the access token'. What do you think?

> - **Config.global_config.jwt_expire**: The period after which the JWT token provided by Taipy will be expired.

Reviewer 2: jwt_expire -> expiration_period

Author: this will result in changing the code in taipy-enterprise, I think it's best to leave it for now and have a different ticket for it. What do you think?

Reviewer: I don't like `jwt_expire` cause it exposes the implementation.
I don't like `expiration_period` either because we don't know that it is related to authentication. 

I would propose `auth_session_duration`.

Once it is released, it will be more difficult to change it. I propose we do the change now in the documentation and I am creating a ticket for renaming the attribute in the code.

Reviewer 2: I don't really mind. My point is that JWT is an implementation detail that users don't care to be exposed to.

Author: as discussed, we have removed this detail from the doc :D

- **Config.global_config.auth_enabled**: The field to enable or disable Taipy authentication and authorization. Its default value is `True`, meaning that the enterprise version, by default, will have authentication and authorization feature enabled.

> Example of configuring LDAP server connection in Taipy:

Reviewer 2: "Example"
->
"Here is an example"

"\`True\`" -> "True"

> ```
> Config.global_config.ldap_server = "ldap://0.0.0.0"
> Config.global_config.ldap_base_dn = "dc=example,dc=org"
> Config.global_config.jwt_secret = "my-32-character-ultra-secure-and-ultra-long-secret"
> Config.global_config.jwt_expire = 600   #for 10 minutes
> Config.global_config.auth_enabled = True
> ```

Reviewer: ```suggestion

=== "Python configuration"

    ```
    Config.configure_global_app(ldap_server="ldap://0.0.0.0", 
                                ldap_base_dn="dc=example,dc=org",
                                jwt_secret = "my-ultra-secure-and-ultra-long-secret",
                                jwt_expire = 600   #for 10 minutes,
                                auth_enabled = True)
    ```

=== "TOML configuration"

    ```python linenums="1"
    from taipy import Config

    Config.load("config.toml")
    ```

    ```toml linenums="1" title="config.toml"

    [TAIPY]

    ldap_server="ldap://0.0.0.0"
    ldap_base_dn="dc=example,dc=org"
    jwt_secret = "my-ultra-secure-and-ultra-long-secret"
    jwt_expire = 600   #for 10 minutes
    auth_enabled = True
    ```
```

> ## Roles

Reviewer 2: -> Predefined roles

"#for 10 minutes" -> "# 600 seconds is 10 minutes"

"Config.global_config.auth_enabled = True": Is that the default?
If not why not?

Author: yes it is the default :D

> ## Roles

Reviewer: ```suggestion
# Roles
```

> Taipy has 4 roles to define a user. Each with different capabilities and will be described in detail below.

Reviewer 2: "Taipy has 4 roles to define a user"
->
"Taipy defines four predefined role names that can be assigned to users"

Reviewer 2: "Each with different"
->
"Each of these predefined roles provide a different set of"

"detail" -> "details"

> Taipy has 4 roles to define a user. Each with different capabilities and will be described in detail below.

Reviewer: ```suggestion
Taipy has four predefined user roles. Each with different capabilities and will be described in detail below.
```

> 

Reviewer 2: In the rest of this document, it's unclear what terms you are using:

Taipy.READER ?
TAIPY_READER ?

What are they, what should be used?
(and of course same for the three other roles)

Author: they are the name of the Taipy roles, it will convert `Taipy.READER` to `TAIPY_READER` for all roles for consistency and to reduce unexplained terms :D

Reviewer 2: I get it (and I can see that in the code).
My point is that this is not clear for users...

Author: I believe these are the roles that will be assigned to the users stored in the ldap server, I don't have a better way to describe this part in the doc actually. @jrobinAV what do you think?

Reviewer: In the roles section, we can specify that the roles (TAIPY_READER, TAIPY_EDITOR, TAIPY_EXECUTOR, TAIPY_ADMIN) must be assigned to the user in the ldap server as a group.

> ### Taipy.READER

Reviewer: ```suggestion
## Taipy.READER
```

> - A reader (TAIPY_READER) should have the right to view the various entities, and related data in a data nodes.

Reviewer 2: Why 'should'?

"the right" -> "the permission"

Drop the comma.

"data nodes" -> "data node"

> - A reader, however, cannot create, edit, delete an entity, or edit the data in a data node manually. They also cannot execute tasks, pipelines or scenarios, subscribe/unsubscribe to an execution, and cancel, pause or resume any jobs.

Reviewer 2: "entity, or" -> "entity or"

Reviewer 2: "They also cannot execute"
->
"Readers are not permitted to submit"

"subscribe/unsubscribe" -> "subscribe or unsubscribe"

", and" -> "or"

"any jobs" -> "any job"

> ### Taipy.EDITOR

Reviewer: ```suggestion
## Taipy.EDITOR
```

> - An editor (TAIPY_EDITOR) should have the right to view, create, edit or delete the various entities, and related data in data nodes.

Reviewer 2: Why 'should'?

"the right" -> "the permission" 

"edit or delete" -> "edit and delete"

Author: Ah yes 'should' is an inappropriate word, I'll change it to 'has'

> - An editor, however, cannot execute any tasks, pipelines or scenarios, subscribe/unsubscribe to an execution, and cancel, pause or resume any jobs.

Reviewer 2: "any tasks, pipelines or scenarios, subscribe/unsubscribe to an execution, and cancel, pause or resume any jobs"
->
"any task, pipeline or scenario, subscribe or unsubscribe to an execution or cancel, pause or resume any job"

> ### Taipy.EXECUTIONER

Reviewer: ```suggestion
## Taipy.EXECUTIONER
```

> - An executor (TAIPY_EXECUTOR) should have the right to view the various entities and related data in data nodes, submit tasks, pipelines or scenarios for execution, subscribe/unsubscribe an execution, and cancel, pause or resume any jobs.

Reviewer 2: Why 'should'?

"the right" -> "the permission"

"subscribe/unsubscribe" -> "subscribe or unsubscribe"

"resume any jobs" -> "resume any job"

> - An executor, however, cannot create, edit or delete the various entities, and related data in data nodes.

Reviewer 2: "entities, and" -> "entities and"

> ### TAIPY.ADMIN

Reviewer: ```suggestion
## TAIPY.ADMIN
```

> - If the authorization feature is activated, a TAIPY_ADMIN should not be restricted at all. He/she should be able to perform all actions available to other roles with no restriction.

Reviewer 2: Why 'should'?

"He/she should be able" -> "A TAIPY_ADMIN will be able"

"restriction" -> "restrictions"

> ## UI page access

Reviewer 2: just add "TODO"
Thx

> ## UI page access

Reviewer: ```suggestion
# UI page access
```

!!! warning "Reminder: Authorization and Authentication is only available in the Enterprise version"

# Pre-requisites

> An LDAP server must be set up by the user. Taipy provides no support for setting up the server.

Reviewer 2: Now that would be:
"If you plan to use LDAP authentication, "...

> An LDAP server must be set up by the user. Taipy provides no support for setting up the server.

Reviewer 2: Now that would be:
"If you plan to use LDAP authentication, "...

# Configuration

> You must configure how Taipy connects to the LDAP server using the Taipy `Config^` features. Here are the configurable properties you can use:

Reviewer 2: Same here: that's true only if LDAP is used.

There should be a 'authentication_protocol' configuration option.

That can be one of 'none', 'ldap', or 'taipy'.

I suspect that 'auth_enabled' should be dropped (implemented by authentication_protocol == None/"none")

Author: @FabienLelaquais would it be better to still have auth_enabled and authentication_protocol having a default value, or just return an error if no protocol is provided. I think it will be clearer that way

Reviewer 2: I would recommend that the default value for 'authentication_protocol ' be None (the user does nothing).
That implies that auth_enabled is False.

That's because, if the users does want authorization, there is configuration to be done.

Author: okay sure, I agree, I'm okay with it :D 

@jrobinAV can I confirm this again with you?

Reviewer: I am ok with this. We need to change that in core/enterprise

Reviewer: ticket created here for taipy-enterprise: https://app.zenhub.com/workspaces/taipy-613ef328affde6000ff77fb3/issues/avaiga/taipy-enterprise/42

Here is an example of configuring LDAP server connection in Taipy:

=== "Python configuration"

    ```
    Config.configure_global_app(ldap_server="ldap://0.0.0.0",
                                ldap_base_dn="dc=example,dc=org",
                                secret_key = "my-ultra-secure-and-ultra-long-secret",
                                auth_session_duration = 600   # 60 seconds is 10 minutes,
                                auth_enabled = True)
    ```

=== "TOML configuration"

    ```python linenums="1"
    from taipy import Config

    Config.load("config.toml")
    ```

    ```toml linenums="1" title="config.toml"

    [TAIPY]

    ldap_server="ldap://0.0.0.0"
    ldap_base_dn="dc=example,dc=org"
    secret_key = "my-ultra-secure-and-ultra-long-secret"
    auth_session_duration = 600   # 60 seconds is 10 minutes,
    auth_enabled = True
    ```

# Roles

Taipy has four predefined user roles names that can be assigned to users. Each of these predefined roles provide a different set of capabilities and will be described in details below.

> !!! warning "Caution: Please note that Taipy does not provide functions to populate the predifined user roles to the user's LDAP server. The user is expected to populate the roles themselves."

Author: @FabienLelaquais I have added a small warning to let the user know about populating the roles to the LDAP and I would definitely like to hear your opinion on it :D

Reviewer 2: That's neat.

The warning box however should be reformatted a little bit (typo fixed in 'predefined')
```
!!! warning
    Caution: Please note that Taipy does not provide functions to populate the predefined user roles to the user's
   LDAP server. The user is expected to populate the roles themselves."
```

- **Config.global_config.ldap_server**: The url of the LDAP server.
- **Config.global_config.ldap_base_dn**: The base DN that the LDAP server uses to search for users authentication.
- **Config.global_config.secret_key**: The secret key that will be used to generate the access token

> - **Config.global_config.auth_session_duration**: The period after which the token provided by Taipy will be expired.

Reviewer 2: => "The period, in seconds, ..."

"will be expired" -> "will expire"

- **Config.global_config.ldap_base_dn**: The base DN that the LDAP server uses to search for users authentication.
- **Config.global_config.secret_key**: The secret key that will be used to generate the access token.
- **Config.global_config.auth_session_duration**: The period after which the token provided by Taipy will expire.

> - **Config.global_config.authenticator_protocol**: The authentication protocol that will be used to authenticate users. Currently Taipy accepts three protocols: None, "ldap" and "taipy" (for testing purposes only). By providing the field with None value, we also disable Taipy authentication and authorization feature. And by choosing other protocols ("ldap" or "taipy"), we also enable Taipy authentication and authorization. The default value of this field is None, meaning that authentication and authorization will be disabled by default.

Reviewer: I would use auth_protocol as well for consistency reasons.
In any case, make sure it is consistent with the code.

Author: sure I completely agree with that :D

    ldap_server="ldap://0.0.0.0"
    ldap_base_dn="dc=example,dc=org"
    secret_key = "my-ultra-secure-and-ultra-long-secret"

>     auth_session_duration = 600   # 60 seconds is 10 minutes,

Reviewer 2: Please remove the trailing comma

    auth_enabled = True
    ```

# Roles

Taipy has four predefined user roles names that can be assigned to users. Each of these predefined roles provide a different set of capabilities and will be described in details below.

> ## TAIPY_READER

Reviewer 2: If all those 'TAIPY_role' are string, I guess you should document them as such:

```## "TAIPY_READER"```
as well as everywhere in the doc content.
